src/app/core/permissions/permission-guard/entity-permission.guard.ts

```diff
@@ -0,0 +1,42 @@
+import { Injectable } from "@angular/core";
+import { ActivatedRouteSnapshot, CanActivate, Router } from "@angular/router";
+import { RouteData } from "../../config/dynamic-routing/view-config.interface";
+import { EntityAbility } from "../ability/entity-ability";
+
+/**
+ * A guard that checks the current users permission to interact with the entity of the route.
+ * Define `requiredPermissionOperation` in the route data / config, to enable a check that will find the relevant entity from config.
+ */
+@Injectable()
+export class EntityPermissionGuard implements CanActivate {
+  constructor(
+    private router: Router,
+    private ability: EntityAbility,
+  ) {}
+
+  canActivate(route: ActivatedRouteSnapshot): boolean {
+    const routeData: RouteData = route.data;
+    if (this.canAccessRoute(routeData)) {
+      return true;
+    } else {
+      if (route instanceof ActivatedRouteSnapshot) {
+        // Route should only change if this is a "real" navigation check (not the check in the NavigationComponent)
+        this.router.navigate(["/404"]);
+      }
+      return false;
+    }
+  }
+
+  private canAccessRoute(routeData) {
+    const operation = routeData?.["requiredPermissionOperation"];
```

Comment:
Hmm what about making this behaviour default? If a "entity" or "config.entity" exists then the "read" permission for this entity is checked. Otherwise its ignored. This would make most route permissions automatic without anything required in the config.

Reply:
Ok but I see the issue that sometimes its not just about reading. I just wonder, whether we can get this a bit more automated because you will always need the same permissions for a certain config.

Reply:
Yeah I think the `RouterService` is the right place for that. In my mind I would have the following cases:

1. If entity can be read from config and a operation is defined check this
2. If entity can be read from config but no operation defined then check with "read"
3. If no entity can be found return true

Potentially we could also save on the component which permission is required (e.g. through a annotation) but thats not possible with the current architecture.